File: 백준/Silver/15650. N과 M （2）/N과 M （2）.cc
#include <iostream>

using namespace std;

int arr[8];

int N, M;

void print(int m) {
	if (m == M) {
		for (int j = 0; j < M; j++) {
			cout << arr[j] << " ";
		}
		cout << endl;
	}
	else {
		for (int i = 1; i <= N; i++) {
			if (m != 0) {
				if (arr[m - 1] < i) {
					arr[m] = i;
					print(m + 1);
					arr[m] = 0;
				}
			}
			else {
				arr[m] = i;
				print(m + 1);
				arr[m] = 0;
			}

		}
	}
}

int main() {
	cin >> N >> M;

	print(0);
}

File: 백준/Silver/15650. N과 M （2）/N과 M （2）.java
import java.util.*;
import java.io.*;

public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
	static StringTokenizer st;
	static int N, M;
	static ArrayList<int[]> result = new ArrayList<>();

	public static void main(String[] args) throws IOException {
		st = new StringTokenizer(br.readLine());
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());

		combination(new int[M], 1, 0);
		for (int[] r : result) {
			for (int i : r) {
				bw.write(i + " ");
			}
			bw.newLine();
		}
		
		bw.close();
	}

	static void combination(int[] arr, int start, int depth) {
		if (depth == M) {
			result.add(arr.clone());
			return;
		}

		for (int i = start; i <= N; i++) {
			arr[depth] = i;
			combination(arr, i + 1, depth + 1);
		}
	}
}